### ISSUE TITLE
Rollup: No-batch-dim support for torch.nn modules

### ISSUE THREAD

**Issue by jbschlosser** [pytorch/pytorch#60585]
## Background
A previous issue (https://github.com/pytorch/pytorch/issues/47149) requested support of arbitrary batch dimensions across module inputs. This rollup issue details the `torch.nn` module updates required to address a subset of this functionality: specifically, the case of no batch dimensions. This particular case is useful for composability with a future vmap.

## Module support
- [x] `nn.AdaptiveAvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveAvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveAvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveLogSoftmaxWithLoss` - Should be updated to support 1D input and scalar target / output1 (@george-qi)
- [x] `nn.AdaptiveMaxPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveMaxPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveMaxPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AlphaDropout`
- [x] `nn.AvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.BCELoss`
- [x] `nn.BCEWithLogitsLoss`
- [x] `nn.Bilinear` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.CELU`
- [x] `nn.CTCLoss` - Should be updated to support a single (source, target) sequence (@george-qi)
- [x] `nn.ConstantPad1d`
- [x] `nn.ConstantPad2d`
- [x] `nn.ConstantPad3d`
- [x] `nn.Conv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.Conv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.Conv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.ConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.ConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.ConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.CosineEmbeddingLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.CosineSimilarity` (pass `dim=0` for no-batch-dim support)
- [x] `nn.CrossEntropyLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.Dropout`
- [x] `nn.Dropout2d`
- [x] `nn.Dropout3d`
- [x] `nn.ELU`
- [x] `nn.Embedding`
- [x] `nn.FeatureAlphaDropout` - Already supports no batch dims, but docs are missing
- [x] `nn.Flatten` (pass `start_dim=0` for no-batch-dim support)
- [x] `nn.Fold` - Should be updated to support 3D inputs (@kshitij12345)
- [x] `nn.FractionalMaxPool2d` - Should be updated to support 3D inputs
- [x] `nn.FractionalMaxPool3d` - Should be updated to support 4D inputs (@george-qi)
- [x] `nn.GELU`
- [x] `nn.GLU`
- [x] `nn.GRU` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.GRUCell` - Should be updated to support 1D input / hidden state (@kshitij12345)
- [x] `nn.GaussianNLLLoss` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.Hardshrink`
- [x] `nn.Hardsigmoid`
- [x] `nn.Hardswish`
- [x] `nn.Hardtanh`
- [x] `nn.HingeEmbeddingLoss`
- [x] `nn.HuberLoss`
- [x] `nn.Identity`
- [x] `nn.InstanceNorm1d` - Should be updated to support 1D or 2D inputs (@kshitij12345)
- [x] `nn.InstanceNorm2d` - Should be updated to support 2D or 3D inputs (@kshitij12345)
- [x] `nn.InstanceNorm3d` - Should be updated to support 3D or 4D inputs (@kshitij12345)
- [x] `nn.KLDivLoss`
- [x] `nn.L1Loss`
- [x] `nn.LPPool1d` - Should be updated to support 2D inputs
- [x] `nn.LPPool2d`
- [x] `nn.LSTM` - Should be updated to support single input, hidden state, cell state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.LSTMCell` - Should be updated to support 1D input, hidden state, cell state (@kshitij12345)
- [x] `nn.LayerNorm` (pass correctly-set `normalized_shape` for no-batch-dim support)
- [x] `nn.LazyConv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyLinear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LeakyReLU`
- [x] `nn.Linear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LogSigmoid`
- [x] `nn.LogSoftmax`
- [x] `nn.MSELoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MarginRankingLoss` - Should be updated to support scalars (@kshitij12345)
- [x] `nn.MaxPool1d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool2d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool3d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxUnpool1d` - Should be updated to support 2D input / indices
- [x] `nn.MaxUnpool2d` - Should be updated to support 3D input / indices
- [x] `nn.MaxUnpool3d` - Should be updated to support 4D input / indices
- [x] `nn.Mish`
- [x] `nn.MultiLabelMarginLoss`
- [x] `nn.MultiLabelSoftMarginLoss` - Should be updated to support 1D input / target
- [x] `nn.MultiMarginLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MultiheadAttention` - Should be updated to support unbatched query, key, values, and masks; `batch_first` arg is meaningless in this case
- [x] `nn.NLLLoss` - Should be updated to support 1D input and scalar target
- [x] `nn.PairwiseDistance` - Should be updated to support 1D inputs with scalar output
- [x] `nn.PixelShuffle`
- [x] `nn.PixelUnshuffle`
- [x] `nn.PoissonNLLLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.RNN` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.RNNCell` - Should be updated to support 1D input and hidden state (@kshitij12345)
- [x] `nn.RReLU`
- [x] `nn.ReLU`
- [x] `nn.ReLU6`
- [x] `nn.ReflectionPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReflectionPad2d`
- [x] `nn.ReflectionPad3d`
- [x] `nn.ReplicationPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReplicationPad2d`
- [x] `nn.ReplicationPad3d`
- [x] `nn.SELU`
- [x] `nn.SiLU`
- [x] `nn.Sigmoid`
- [x] `nn.SmoothL1Loss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.SoftMarginLoss`
- [x] `nn.Softmax`
- [x] `nn.Softmax2d` - Should be updated to support 3D input
- [x] `nn.Softmin`
- [x] `nn.Softplus`
- [x] `nn.Softshrink`
- [x] `nn.Softsign`
- [x] `nn.Tanh`
- [x] `nn.Tanhshrink`
- [x] `nn.Threshold`
- [x] `nn.Transformer` - Should be updated to support unbatched source / target sequences and masks (@kshitij12345)
- [x] `nn.TransformerDecoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TransformerEncoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TripletMarginLoss` - Should be updated to support 1D anchor, positive, and negative (@kshitij12345)
- [x] `nn.TripletMarginWithDistanceLoss` (support is based on the choice of distance function)
- [x] `nn.Unflatten` - Already supports no batch dims, but docs should make this clear
- [x] `nn.ZeroPad2d`

## Semantically incompatible modules
* `nn.BatchNorm1d` - Only defined over batch
* `nn.BatchNorm2d` - Only defined over batch
* `nn.BatchNorm3d` - Only defined over batch
* `nn.LazyBatchNorm1d` - Only defined over batch
* `nn.LazyBatchNorm2d` - Only defined over batch
* `nn.LazyBatchNorm3d` - Only defined over batch
* `nn.SyncBatchNorm` - Only defined over batch

## Modules that require BC-breaking changes
* `nn.ChannelShuffle` - While the docs say that the module accepts shape (*, C, H, W), the implementation assumes shape (N, C, *) with * being 1 or more dimensions; switching to (C, *) is BC-breaking because it would reinterpret dims
* `nn.EmbeddingBag` - Already supports 1D inputs with different semantics
* `nn.GroupNorm` - Supports arbitrary spatial dims; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.LocalResponseNorm` - Supports (N, C, *) shape now; switching to (C, *) is BC-breaking since it would reinterpret dims
* `nn.PReLU` - Always assumes the 2nd dim is the channels dim
* `nn.Unfold` - Supports 4D inputs; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.Upsample` - Supporting unbatched inputs would be BC-breaking because it would reinterpret dims

## Irrelevant modules
* `nn.Module`
* `nn.ModuleDict`
* `nn.ModuleList`
* `nn.ParameterDict`
* `nn.ParameterList`
* `nn.Sequential`
* `nn.TransformerDecoder` - see `nn.TransformerDecoderLayer` instead
* `nn.TransformerEncoder` - see `nn.TransformerEncoderLayer` instead

cc @albanD @mruberry @jbschlosser

**Comment by thomasjpfan** [comment]
What does the semantics of no-batch-dim for criterion look like when there is a reduction? Consider:

```python
import torch
import torch.nn as nn

loss_mean = nn.MSELoss(reduction='mean')
input = torch.randn(3, 5, 5, requires_grad=True)
target = torch.randn(3, 5, 5)

torch.isclose(
    loss_mean(input, target),
    # Need to scale the sum down by the number of batches
    sum(loss_mean(input[i], target[i]) for i in range(3)) / 3
)
```

**Comment by jbschlosser** [comment]
> What does the semantics of no-batch-dim for criterion look like when there is a reduction? Consider:

Great question - I'd say that "reduction" only has non-trivial meaning when being applied over a batch. If there's only a single item in an implicit "batch" (i.e. the no batch dim case), I don't see a problem with the output being equivalent across all reduction types.

Note that you could tweak your example to pass 3 inputs of shape `(1, 5, 5)` and you'd run into the same thing.

**Comment by mruberry** [comment]
This looks like it was accidentally closed?

**Comment by jbschlosser** [comment]
Quick update: for modules where the input is ambiguous (e.g. it could be either a 3D input with no batch dim or a 2D input with a batch dim), there are a few options we've discussed:
1. Add more `nn.Module*d`s to ensure that inputs are always unambiguous.
    - This could apply to `nn.ChannelShuffle`, `nn.GroupNorm`, `nn.Unfold`, and `nn.Upsample` above.
    - Notably, this approach doesn't address the functional forms (e.g. `nn.InstanceNorm*d` exist but they all call into `F.instance_norm`), but we could also add e.g. `F.instance_norm*d`.
    - This fits well with what we already provide, but goes against the request in #41243.
2. Add a new namespace for modules / functions that operate on no-batch-dim inputs. This would keep them separate from the ones we already have and it would be unambiguous what the user means when those are used.
    - Lack of discoverability and high maintenance costs for a new set of things within a new namespace may be downsides here.
3. Provide an arg like `num_spatial_dims`, with any inputs dims beyond the number specified being interpreted as batch dims.
    - This is what various libraries in JAX (Elegy, Haiku) do for convolution.
    - Generally, conv*d layers are provided on top of a general conv that takes `num_spatial_dims`, so this composes well with (1).
4. Provide an arg like `axis` / `dims`, indicating which dims to operate over.
    - This is what various libraries in JAX (Objax, Trax, Haiku) do for normalization.
    - I think it's a little bit strange to provide BatchNorm, InstanceNorm, etc. separately but still make the axes of operation configurable. Performing a BatchNorm with some sets of axes is no longer truly a BatchNorm, for example.
    - It's also arguably an awkward UX to require providing a set of dims just to handle no-batch-dim input.

Based on the above, I'm personally leaning towards (3). In my mind, it would be nice to have a single functional form that handles all input shapes, based on the value of `num_spatial_dims`. The various `Module*d` / `F.function_*d` forms that exist could call into this, keeping the same UX as now, but longer term we'd still have the possibility of removing these if we want to satisfy #41243.

**Comment by jbschlosser** [comment]
Wrongly closed :)

**Comment by kshitij12345** [comment]
Oops! Had incorrectly marked the PR with Fixes
